File: src/experiment/governance.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import { queryExperimentResults } from "./results.js";
import type { VariantResult } from "./results.js";
import type { FactorDelta } from "../factors/snapshots.js";

export type ThresholdOperator = "gt" | "lt" | "gte" | "lte" | "eq";

export interface Threshold {
  id: string;
  factor_name: string;
  component_path: string | null;
  operator: ThresholdOperator;
  value: number;
  action: "alert" | "experiment";
}

export interface FactorVerdict {
  factor_name: string;
  best_variant: string;
  best_delta: number;
  control_delta: number;
  is_significant: boolean;
}

export type GovernanceAction = "conclude" | "flag_review" | "continue";

export interface GovernanceVerdict {
  action: GovernanceAction;
  winning_variant: string | null;
  factor_verdicts: FactorVerdict[];
}

/**
 * Orchestrator: evaluates whether an experiment should be concluded
 * based on factor deltas exceeding governance thresholds.
 */
export async function evaluateExperimentThresholds(
  client: SupabaseClient,
  experimentId: string,
  factorNames: string[],
): Promise<GovernanceVerdict> {
  const componentPath = await fetchExperimentComponentPath(client, experimentId);
  if (!componentPath) return buildEmptyVerdict();

  const results = await queryExperimentResults(client, experimentId, factorNames);
  const thresholds = await fetchThresholds(client, factorNames, componentPath);

  return computeGovernanceVerdict(results, thresholds);
}

/**
 * Concludes an experiment: sets status to 'concluded' and records the winner.
 */
export async function concludeExperiment(
  client: SupabaseClient,
  experimentId: string,
  winningVariant: string,
): Promise<void> {
  const { error } = await client
    .from("experiments")
    .update({
      status: "concluded",
      concluded_at: new Date().toISOString(),
      winning_variant: winningVariant,
    })
    .eq("id", experimentId)
    .eq("status", "running");

  if (error) throw new Error(`concludeExperiment failed: ${error.message}`);
}

async function fetchExperimentComponentPath(
  client: SupabaseClient,
  experimentId: string,
): Promise<string | null> {
  const { data, error } = await client
    .from("experiments")
    .select("component_path")
    .eq("id", experimentId)
    .maybeSingle();

  if (error || !data) return null;
  return data.component_path as string;
}

async function fetchThresholds(
  client: SupabaseClient,
  factorNames: string[],
  componentPath: string,
): Promise<Threshold[]> {
  const { data, error } = await client
    .from("thresholds")
    .select("id, factor_name, component_path, operator, value, action")
    .in("factor_name", factorNames)
    .or(`component_path.eq.${componentPath},component_path.is.null`);

  if (error) throw new Error(`fetchThresholds failed: ${error.message}`);
  return (data ?? []) as Threshold[];
}

/**
 * Pure logic: compares variant deltas against thresholds to produce a verdict.
 * SQL mirror: auxi.evaluate_governance() in migration 021 duplicates these rules.
 *
 * Conservative rules:
 * - At least one factor must have a significant delta to conclude
 * - All significant factors must agree on the same winning variant
 * - If any treatment delta is worse than control, flag for review
 * - Otherwise, continue running
 */
export function computeGovernanceVerdict(
  results: VariantResult[],
  thresholds: Threshold[],
): GovernanceVerdict {
  if (results.length < 2 || thresholds.length === 0) return buildEmptyVerdict();

  const controlResult = results.find(r => r.variant_key === "control");
  if (!controlResult) return buildEmptyVerdict();

  const treatmentResults = results.filter(r => r.variant_key !== "control");
  const factorVerdicts = buildFactorVerdicts(controlResult, treatmentResults, thresholds);

  return deriveVerdict(factorVerdicts);
}

function buildFactorVerdicts(
  controlResult: VariantResult,
  treatmentResults: VariantResult[],
  thresholds: Threshold[],
): FactorVerdict[] {
  const verdicts: FactorVerdict[] = [];

  for (const threshold of thresholds) {
    const controlDelta = findFactorDelta(controlResult.factor_deltas, threshold.factor_name);
    const bestTreatment = findBestTreatment(treatmentResults, threshold);

    if (!bestTreatment || controlDelta === null) continue;

    const improvement = Math.abs(bestTreatment.delta) - Math.abs(controlDelta);
    const isSignificant = isThresholdExceeded(improvement, threshold.operator, threshold.value);

    verdicts.push({
      factor_name: threshold.factor_name,
      best_variant: bestTreatment.variantKey,
      best_delta: bestTreatment.delta,
      control_delta: controlDelta,
      is_significant: isSignificant,
    });
  }

  return verdicts;
}

interface TreatmentCandidate {
  variantKey: string;
  delta: number;
}

function findBestTreatment(
  treatmentResults: VariantResult[],
  threshold: Threshold,
): TreatmentCandidate | null {
  let best: TreatmentCandidate | null = null;

  for (const result of treatmentResults) {
    const delta = findFactorDelta(result.factor_deltas, threshold.factor_name);
    if (delta === null) continue;

    if (!best || Math.abs(delta) > Math.abs(best.delta)) {
      best = { variantKey: result.variant_key, delta };
    }
  }

  return best;
}

function findFactorDelta(deltas: FactorDelta[], factorName: string): number | null {
  const match = deltas.find(d => d.factor_name === factorName);
  return match?.delta ?? null;
}

export function isThresholdExceeded(
  value: number,
  operator: ThresholdOperator,
  threshold: number,
): boolean {
  switch (operator) {
    case "gt": return value > threshold;
    case "lt": return value < threshold;
    case "gte": return value >= threshold;
    case "lte": return value <= threshold;
    case "eq": return value === threshold;
  }
}

function deriveVerdict(factorVerdicts: FactorVerdict[]): GovernanceVerdict {
  if (factorVerdicts.length === 0) return buildEmptyVerdict();

  const hasWorseningTreatment = factorVerdicts.some(
    v => Math.abs(v.best_delta) < Math.abs(v.control_delta),
  );
  if (hasWorseningTreatment) {
    return { action: "flag_review", winning_variant: null, factor_verdicts: factorVerdicts };
  }

  const significantVerdicts = factorVerdicts.filter(v => v.is_significant);
  if (significantVerdicts.length === 0) {
    return { action: "continue", winning_variant: null, factor_verdicts: factorVerdicts };
  }

  const winningVariants = new Set(significantVerdicts.map(v => v.best_variant));
  if (winningVariants.size !== 1) {
    return { action: "flag_review", winning_variant: null, factor_verdicts: factorVerdicts };
  }

  const winningVariant = [...winningVariants][0];
  return { action: "conclude", winning_variant: winningVariant, factor_verdicts: factorVerdicts };
}

function buildEmptyVerdict(): GovernanceVerdict {
  return { action: "continue", winning_variant: null, factor_verdicts: [] };
}

File: packages/kotlin-engine/src/commonMain/kotlin/ai/factoredui/engine/experiments/Governance.kt
package ai.factoredui.engine.experiments

import ai.factoredui.engine.factors.FactorDelta
import kotlin.math.abs

/**
 * Governance verdict logic — the gate that decides whether an experiment may
 * conclude, must be flagged for human review, or should keep running.
 *
 * Ported from `packages/core/src/experiment/governance.ts`. Pure. Conservative
 * rules:
 * - Need ≥2 variants, ≥1 threshold, and a `control` variant, else continue.
 * - A treatment whose magnitude is worse than control on any factor ⇒ flag_review.
 * - No significant factor ⇒ continue.
 * - Significant factors disagreeing on the winner ⇒ flag_review.
 * - Exactly one winner ⇒ conclude.
 */

private const val CONTROL_VARIANT = "control"

fun computeGovernanceVerdict(
    results: List<VariantResult>,
    thresholds: List<Threshold>,
): GovernanceVerdict {
    if (results.size < 2 || thresholds.isEmpty()) return emptyVerdict()

    val controlResult = results.firstOrNull { it.variantKey == CONTROL_VARIANT } ?: return emptyVerdict()
    val treatmentResults = results.filter { it.variantKey != CONTROL_VARIANT }
    val factorVerdicts = buildFactorVerdicts(controlResult, treatmentResults, thresholds)

    return deriveVerdict(factorVerdicts)
}

fun isThresholdExceeded(value: Double, operator: ThresholdOperator, threshold: Double): Boolean =
    when (operator) {
        ThresholdOperator.GT -> value > threshold
        ThresholdOperator.LT -> value < threshold
        ThresholdOperator.GTE -> value >= threshold
        ThresholdOperator.LTE -> value <= threshold
        ThresholdOperator.EQ -> value == threshold
    }

private fun buildFactorVerdicts(
    controlResult: VariantResult,
    treatmentResults: List<VariantResult>,
    thresholds: List<Threshold>,
): List<FactorVerdict> {
    val verdicts = mutableListOf<FactorVerdict>()

    for (threshold in thresholds) {
        val controlDelta = findFactorDelta(controlResult.factorDeltas, threshold.factorName)
        val bestTreatment = findBestTreatment(treatmentResults, threshold)
        if (bestTreatment == null || controlDelta == null) continue

        val improvement = abs(bestTreatment.delta) - abs(controlDelta)
        val isSignificant = isThresholdExceeded(improvement, threshold.operator, threshold.value)

        verdicts.add(
            FactorVerdict(
                factorName = threshold.factorName,
                bestVariant = bestTreatment.variantKey,
                bestDelta = bestTreatment.delta,
                controlDelta = controlDelta,
                isSignificant = isSignificant,
            ),
        )
    }

    return verdicts
}

private data class TreatmentCandidate(val variantKey: String, val delta: Double)

private fun findBestTreatment(
    treatmentResults: List<VariantResult>,
    threshold: Threshold,
): TreatmentCandidate? {
    var best: TreatmentCandidate? = null
    for (result in treatmentResults) {
        val delta = findFactorDelta(result.factorDeltas, threshold.factorName) ?: continue
        if (best == null || abs(delta) > abs(best.delta)) {
            best = TreatmentCandidate(result.variantKey, delta)
        }
    }
    return best
}

private fun findFactorDelta(deltas: List<FactorDelta>, factorName: String): Double? =
    deltas.firstOrNull { it.factorName == factorName }?.delta

private fun deriveVerdict(factorVerdicts: List<FactorVerdict>): GovernanceVerdict {
    if (factorVerdicts.isEmpty()) return emptyVerdict()

    val hasWorseningTreatment = factorVerdicts.any { abs(it.bestDelta) < abs(it.controlDelta) }
    if (hasWorseningTreatment) {
        return GovernanceVerdict(GovernanceAction.FLAG_REVIEW, null, factorVerdicts)
    }

    val significantVerdicts = factorVerdicts.filter { it.isSignificant }
    if (significantVerdicts.isEmpty()) {
        return GovernanceVerdict(GovernanceAction.CONTINUE, null, factorVerdicts)
    }

    val winningVariants = significantVerdicts.map { it.bestVariant }.toSet()
    if (winningVariants.size != 1) {
        return GovernanceVerdict(GovernanceAction.FLAG_REVIEW, null, factorVerdicts)
    }

    return GovernanceVerdict(GovernanceAction.CONCLUDE, winningVariants.first(), factorVerdicts)
}

private fun emptyVerdict(): GovernanceVerdict =
    GovernanceVerdict(GovernanceAction.CONTINUE, null, emptyList())
